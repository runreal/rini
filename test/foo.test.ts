import { assertNotEquals } from 'std/assert/mod.ts'
import { assertSnapshot } from 'std/testing/snapshot.ts'
import * as path from 'std/path/mod.ts'
import * as ini from '../mod.ts'

const fixture = path.resolve('./test/fixtures/foo.ini')
const data = Deno.readTextFileSync(fixture)

function serializer(actual: string): string {
	return actual.replace(/\r\n/g, '\n')
}

Deno.test('decode from file', async function (t): Promise<void> {
	const d = ini.decode(data)
	await assertSnapshot(t, d)
})

Deno.test('encode from data', async function (t): Promise<void> {
	const d = ini.decode(data)
	const e = ini.encode(d)
	await assertSnapshot(t, e, { serializer })
})

Deno.test('never a blank first or last line', function () {
	const obj = { log: { type: 'file', level: { label: 'debug', value: 10 } } }
	const e = ini.encode(obj)
	assertNotEquals(e.slice(0, 1), '\n', 'Never a blank first line')
	assertNotEquals(e.slice(-2), '\n\n', 'Never a blank final line')
})

Deno.test('encode with option', async function (t): Promise<void> {
	const obj = { log: { type: 'file', level: { label: 'debug', value: 10 } } }
	const e = ini.encode(obj, { section: 'prefix' })
	await assertSnapshot(t, e, { serializer })
})

Deno.test('encode with whitespace', async function (t): Promise<void> {
	const obj = { log: { type: 'file', level: { label: 'debug', value: 10 } } }
	const e = ini.encode(obj, { whitespace: true })
	await assertSnapshot(t, e, { serializer })
})

Deno.test('encode with newline', async function (t): Promise<void> {
	const obj = { log: { type: 'file', level: { label: 'debug', value: 10 } } }
	const e = ini.encode(obj, { newline: true })
	await assertSnapshot(t, e, { serializer })
})

Deno.test('encode with platform=windows', async function (t): Promise<void> {
	const obj = { log: { type: 'file', level: { label: 'debug', value: 10 } } }
	const e = ini.encode(obj, { platform: 'windows' })
	await assertSnapshot(t, e.split('\r\n'))
})

Deno.test('encode with align', async function (t): Promise<void> {
	const d = ini.decode(data)
	const e = ini.encode(d, { align: true })
	await assertSnapshot(t, e, { serializer })
})

Deno.test('encode with sort', async function (t): Promise<void> {
	const d = ini.decode(data)
	const e = ini.encode(d, { sort: true })
	await assertSnapshot(t, e, { serializer })
})

Deno.test('encode with align and sort', async function (t): Promise<void> {
	const d = ini.decode(data)
	const e = ini.encode(d, { align: true, sort: true })
	await assertSnapshot(t, e, { serializer })
})
