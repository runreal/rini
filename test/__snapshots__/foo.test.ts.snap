export const snapshot = {};

snapshot[`decode from file 1`] = `
[Object: null prototype] {
  " xa  n          p ": '"\r\nyoyoyo\r\r\n',
  "[disturbing]": "hey you never know",
  "a with spaces": "b  c",
  "x.y.z": [Object: null prototype] {
    "a.b.c": [Object: null prototype] {
      "a.b.c": "abc",
      noHashComment: "this# this is not a comment",
      nocomment: "this; this is not a comment",
    },
    "x.y.z": "xyz",
  },
  a: [Object: null prototype] {
    "[]": "a square?",
    av: "a val",
    b: [Object: null prototype] {
      c: [Object: null prototype] {
        e: "1",
        j: "2",
      },
    },
    cr: [
      "four",
      "eight",
    ],
    e: '{ o: p, a: { av: a val, b: { c: { e: "this [value]" } } } }',
    j: '"{ o: "p", a: { av: "a val", b: { c: { e: "this [value]" } } } }"',
  },
  ar: [
    "one",
    "three",
    "this is included",
  ],
  b: [Object: null prototype] {},
  br: "warm",
  eq: "eq=eq",
  false: false,
  null: null,
  o: "p",
  s1: \`"something'\`,
  s2: "something else",
  s3: "",
  s4: "",
  s5: "   ",
  s6: " a ",
  s7: true,
  s: "something",
  true: true,
  undefined: "undefined",
  zr: [
    "deedee",
  ],
}
`;

snapshot[`encode from data 1`] = `
o=p
a with spaces=b  c
" xa  n          p "="\\"\\r\\nyoyoyo\\r\\r\\n"
"[disturbing]"=hey you never know
s=something
s1="something'
s2=something else
s3=
s4=
s5="   "
s6=" a "
s7=true
true=true
false=false
null=null
undefined=undefined
zr[]=deedee
ar[]=one
ar[]=three
ar[]=this is included
br=warm
eq="eq=eq"

[a]
av=a val
e={ o: p, a: { av: a val, b: { c: { e: "this [value]" } } } }
j="\\"{ o: \\"p\\", a: { av: \\"a val\\", b: { c: { e: \\"this [value]\\" } } } }\\""
"[]"=a square?
cr[]=four
cr[]=eight

[a.b.c]
e=1
j=2

[x\\.y\\.z]
x.y.z=xyz

[x\\.y\\.z.a\\.b\\.c]
a.b.c=abc
nocomment=this\\; this is not a comment
noHashComment=this\\# this is not a comment

`

snapshot[`encode with align 1`] = `
o                    = p
a with spaces        = b  c
" xa  n          p " = "\\"\\r\\nyoyoyo\\r\\r\\n"
"[disturbing]"       = hey you never know
s                    = something
s1                   = "something'
s2                   = something else
s3                   = 
s4                   = 
s5                   = "   "
s6                   = " a "
s7                   = true
true                 = true
false                = false
null                 = null
undefined            = undefined
zr[]                 = deedee
ar[]                 = one
ar[]                 = three
ar[]                 = this is included
br                   = warm
eq                   = "eq=eq"

[a]
av   = a val
e    = { o: p, a: { av: a val, b: { c: { e: "this [value]" } } } }
j    = "\\"{ o: \\"p\\", a: { av: \\"a val\\", b: { c: { e: \\"this [value]\\" } } } }\\""
"[]" = a square?
cr[] = four
cr[] = eight

[a.b.c]
e = 1
j = 2

[x\\.y\\.z]
x.y.z = xyz

[x\\.y\\.z.a\\.b\\.c]
a.b.c         = abc
nocomment     = this\\; this is not a comment
noHashComment = this\\# this is not a comment

`

snapshot[`encode with align and sort 1`] = `
" xa  n          p " = "\\"\\r\\nyoyoyo\\r\\r\\n"
"[disturbing]"       = hey you never know
a with spaces        = b  c
ar[]                 = one
ar[]                 = three
ar[]                 = this is included
br                   = warm
eq                   = "eq=eq"
false                = false
null                 = null
o                    = p
s                    = something
s1                   = "something'
s2                   = something else
s3                   = 
s4                   = 
s5                   = "   "
s6                   = " a "
s7                   = true
true                 = true
undefined            = undefined
zr[]                 = deedee

[a]
"[]" = a square?
av   = a val
cr[] = four
cr[] = eight
e    = { o: p, a: { av: a val, b: { c: { e: "this [value]" } } } }
j    = "\\"{ o: \\"p\\", a: { av: \\"a val\\", b: { c: { e: \\"this [value]\\" } } } }\\""

[a.b.c]
e = 1
j = 2

[x\\.y\\.z]
x.y.z = xyz

[x\\.y\\.z.a\\.b\\.c]
a.b.c         = abc
noHashComment = this\\# this is not a comment
nocomment     = this\\; this is not a comment

`

snapshot[`encode with newline 1`] = `
[log]

type=file

[log.level]

label=debug
value=10

`

snapshot[`encode with option 1`] = `
[prefix.log]
type=file

[prefix.log.level]
label=debug
value=10

`

snapshot[`encode with platform=windows 1`] = `
[
  "[log]",
  "type=file",
  "",
  "[log.level]",
  "label=debug",
  "value=10",
  "",
]
`

snapshot[`encode with sort 1`] = `
" xa  n          p "="\\"\\r\\nyoyoyo\\r\\r\\n"
"[disturbing]"=hey you never know
a with spaces=b  c
ar[]=one
ar[]=three
ar[]=this is included
br=warm
eq="eq=eq"
false=false
null=null
o=p
s=something
s1="something'
s2=something else
s3=
s4=
s5="   "
s6=" a "
s7=true
true=true
undefined=undefined
zr[]=deedee

[a]
"[]"=a square?
av=a val
cr[]=four
cr[]=eight
e={ o: p, a: { av: a val, b: { c: { e: "this [value]" } } } }
j="\\"{ o: \\"p\\", a: { av: \\"a val\\", b: { c: { e: \\"this [value]\\" } } } }\\""

[a.b.c]
e=1
j=2

[x\\.y\\.z]
x.y.z=xyz

[x\\.y\\.z.a\\.b\\.c]
a.b.c=abc
noHashComment=this\\# this is not a comment
nocomment=this\\; this is not a comment

`

snapshot[`encode with whitespace 1`] = `
[log]
type = file

[log.level]
label = debug
value = 10

`
